src/filter_tree/filter_tree.c

```diff
@@ -163,33 +72,28 @@ Vector* FilterTree_SubTrees(const FT_FilterNode *root) {
 }
 
 FT_FilterNode* BuildFiltersTree(const AST_FilterNode *root) {
+    FT_FilterNode *filterNode;
+
     if(root->t == N_PRED) {
-        if(root->pn.t == N_CONSTANT) {
-            return _CreateConstFilterNode(root->pn);
-        } else {
-            return _CreateVaryingFilterNode(root->pn);
-        }
-	}
-
-    // root->t == N_COND
-    // Create condition node
-    FT_FilterNode* filterNode = CreateCondFilterNode(root->cn.op);
-    AppendLeftChild(filterNode, BuildFiltersTree(root->cn.left));
-    AppendRightChild(filterNode, BuildFiltersTree(root->cn.right));
-	return filterNode;
+        filterNode = malloc(sizeof(FT_FilterNode));
+        filterNode->t= FT_N_PRED;
+        filterNode->pred.op = root->pn.op;
+        filterNode->pred.lhs = AR_EXP_BuildFromAST(root->pn.lhs, NULL);
+        filterNode->pred.rhs = AR_EXP_BuildFromAST(root->pn.rhs, NULL);
+    } else {
+        filterNode = CreateCondFilterNode(root->cn.op);
+        AppendLeftChild(filterNode, BuildFiltersTree(root->cn.left));
+        AppendRightChild(filterNode, BuildFiltersTree(root->cn.right));
+    }
+
+    return filterNode;
 }
 
 /* Applies a single filter to a single result.
  * Compares given values, tests if values maintain desired relation (op) */
-int _applyFilter(SIValue* aVal, SIValue* bVal, CmpFunc f, int op) {
-    /* TODO: Make sure values type confirms with compare function
-     * TODO: If there's a type mismatch should we try to find a common type and compare ? */
-    if(aVal->type != bVal->type) {
-        // Type mismatch!
-        return 0;
-    }
-
-    int rel = f(aVal, bVal);
+int _applyFilter(SIValue* aVal, SIValue* bVal, int op) {
+    /* TODO Consider updating all logic around comparison routines */
+    int rel = SIValue_Compare(*aVal, *bVal);
```

Comment:
I'd like to do some benchmarking of how much this change costs, change `SIValue_Compare` and its callers so that it doesn't try to work for ordering as well as comparing, and delete `value_cmp.c` and `value_cmp.h`.